applogic/src/util/background.rs

```diff
@@ -0,0 +1,143 @@
+// SPDX-FileCopyrightText: 2025 Phoenix R&D GmbH <hello@phnx.im>
+//
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::{
+    fmt,
+    marker::PhantomData,
+    pin::pin,
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use tokio::time;
+use tokio_stream::{Stream, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{debug, error, info};
+use uuid::Uuid;
+
+use super::{FibonacciBackoff, spawn_from_sync};
+
+/// Timeout after a stream stop is not considered as error
+const REGULAR_STOP_TIMEOUT: Duration = Duration::from_secs(30 * 60 * 60); // 30 minutes
+
+/// A task that runs in the background and handles events from a stream.
+///
+/// The task is created from a [`BackgroundStreamContext`], which is responsible for creating the
+/// stream and handling events. The task is responsible for managing the lifetime of the stream:
+///
+/// - stream is created by the [`BackgroundStreamContext`]
+/// - stream is stopped when the app goes in the background
+/// - stream is restarted when the app goes in the foreground
+/// - stream is restarted on errors with a fibonacci backoff
+/// - stream is restarted when it gracefully stops
+pub(crate) struct BackgroundStreamTask<C, Event> {
+    id: Uuid,
+    name: Arc<str>,
+    context: C,
+    cancel: CancellationToken,
+    backoff: FibonacciBackoff,
+    _marker: PhantomData<Event>,
+}
+
+impl<C, Event> BackgroundStreamTask<C, Event>
+where
+    C: BackgroundStreamContext<Event> + Send + 'static,
+    Event: fmt::Debug + Send + 'static,
+{
+    pub(crate) fn new(name: impl Into<String>, context: C, cancel: CancellationToken) -> Self {
+        let id = Uuid::new_v4();
+        let name = name.into().into();
+        debug!(%name, %id, "new background stream task");
+        Self {
+            id,
+            name,
+            context,
+            cancel,
+            backoff: FibonacciBackoff::new(),
+            _marker: PhantomData,
+        }
+    }
+
+    pub(crate) fn spawn(mut self) {
+        spawn_from_sync(async move {
+            loop {
+                let started_at = Instant::now();
+                let res = self.run().await;
+
+                // stop handler on cancellation
+                if self.cancel.is_cancelled() {
+                    info!(name = %self.name, id = %self.id, "background stream stopped");
+                    return;
+                }
+
+                // wait until the app is in foreground
+                self.context.in_foreground().await;
+
+                if let Err(error) = res {
+                    // if failed faster than regular stop, retry with backoff
+                    let timeout = if started_at.elapsed() < REGULAR_STOP_TIMEOUT {
+                        self.backoff.next_backoff()
+                    } else {
+                        self.backoff.reset();
+                        Duration::default()
```

Comment:
Don't we want to rate-limit this at least a little bit? I know we should normally not end up here, but if we do for some reason (e.g. clock drift), we might just keep running into the same error again and again. I'm thinking a backoff of 1s might defuse this.

Reply:
Technically, this is not a problem because this branch is only hit when the last reconnect was at least after `REGULAR_STOP_TIMEOUT`. However, there is actually a problem in the branch when there is no error. There we have to apply a backoff.

To summarize:

When there is an error or we are under the `REGULAR_STOP_TIMEOUT` => backoff,
When there is no error and we are after the `REGULAR_STOP_TIMEOUT` => backoff is not needed because enough time already elapsed.

This is now fixed in the next commit.